lib/propolis-client/src/instance_spec.rs

```diff
@@ -52,8 +52,52 @@ use serde::{Deserialize, Serialize};
 pub use crucible::VolumeConstructionRequest;
 pub use propolis_types::PciPath;
 
+/// Type alias for keys in the instance spec's maps.
+type SpecKey = String;
+
+/// Routines used to check whether two components are migration-compatible.
+pub trait MigrationCompatible {
+    /// Returns true if `self` and `other` describe spec elements that are
+    /// similar enough to permit migration of this element from one VMM to
+    /// another.
+    ///
+    /// Note that this can be, but isn't always, a simple check for equality.
+    /// Backends, in particular, may be migration-compatible but have different
+    /// configuration payloads. The migration protocol allows components like
+    /// this to augment this check with their own compatibility checks.
+    fn is_migration_compatible(&self, other: &Self) -> bool;
+}
+
+impl<T: MigrationCompatible> MigrationCompatible for BTreeMap<SpecKey, T> {
+    // Two keyed maps of components are compatible if they contain all the same
+    // keys and if, for each key, the corresponding values are
+    // migration-compatible.
+    fn is_migration_compatible(&self, other: &Self) -> bool {
+        // If the two maps have different sizes, then they have different key
+        // sets.
+        if self.len() != other.len() {
+            return false;
+        }
+
+        // Each key in `self`'s map must be present in `other`'s map, and the
+        // corresponding values must be compatible with one another.
+        for (key, this_val) in self.iter() {
+            match other.get(key) {
+                Some(other_val) => {
+                    if !this_val.is_migration_compatible(other_val) {
+                        return false;
+                    }
+                }
+                None => return false,
+            }
+        }
+
+        true
```

Comment:
Since loops vs iter combinators came up in the space today :P

```suggestion
        self.iter().all(|(key, this_val)| {
            other.get(key).map_or(false, |other_val| {
                this_val.is_migration_compatible(other_val)
            })
        })
```

Reply:
WHOA

Reply:
Nifty!